src/abstractdataframe/abstractdataframe.jl

```diff
@@ -2289,6 +2289,277 @@ julia> reverse(df)
+
+`permute!` will produce a correct result even if some columns of passed data frame
+or permutation `p` are identical (checked with `===`). Otherwise, if two columns share
+some part of memory but are not identical (e.g. are different views of the same parent
+vector) then `permute!` result might be incorrect.
+
+# Examples
+julia> df = DataFrame(a=1:5, b=6:10, c=11:15)
+5×3 DataFrame
+ Row │ a      b      c
+     │ Int64  Int64  Int64
+─────┼─────────────────────
+   1 │     1      6     11
+   2 │     2      7     12
+   3 │     3      8     13
+   4 │     4      9     14
+   5 │     5     10     15
+
+julia> permute!(df, [5, 3, 1, 2, 4])
+5×3 DataFrame
+ Row │ a      b      c
+     │ Int64  Int64  Int64
+─────┼─────────────────────
+   1 │     5     10     15
+   2 │     3      8     13
+   3 │     1      6     11
+   4 │     2      7     12
+   5 │     4      9     14
+"""
+function Base.permute!(df::AbstractDataFrame, p::AbstractVector{<:Integer})
+    toskip = Set{Int}()
+    seen_cols = IdDict{Any, Nothing}()
+    for (i, col) in enumerate(eachcol(df))
+        if haskey(seen_cols, col)
+            push!(toskip, i)
+        else
+            seen_cols[col] = nothing
+        end
+        # p might be a column of df so we make sure we unalias
+        # we do not
+        if col === p
+            p = copy(p)
+        end
+    end
+
+    pp = similar(p)
+
+    for (i, col) in enumerate(eachcol(df))
+        if !(i in toskip)
+            copyto!(pp, p)
+            Base.permute!!(col, pp)
+        end
+    end
+    return df
+end
+
+"""
+    invpermute!(df::AbstractDataFrame, p)
+
+Like [`permute!`](@ref), but the inverse of the given permutation is applied.
+
+`invpermute!` will produce a correct result even if some columns of passed data
+frame or permutation `p` are identical (checked with `===`). Otherwise, if two
+columns share some part of memory but are not identical (e.g. are different views
+of the same parent vector) then `permute!` result might be incorrect.
+
+# Examples
+
+julia> df = DataFrame(a=1:5, b=6:10, c=11:15)
+5×3 DataFrame
+ Row │ a      b      c
+     │ Int64  Int64  Int64
+─────┼─────────────────────
+   1 │     1      6     11
+   2 │     2      7     12
+   3 │     3      8     13
+   4 │     4      9     14
+   5 │     5     10     15
+
+julia> permute!(df, [5, 3, 1, 2, 4])
+5×3 DataFrame
+ Row │ a      b      c
+     │ Int64  Int64  Int64
+─────┼─────────────────────
+   1 │     5     10     15
+   2 │     3      8     13
+   3 │     1      6     11
+   4 │     2      7     12
+   5 │     4      9     14
+
+julia> invpermute!(df, [5, 3, 1, 2, 4])
+5×3 DataFrame
+ Row │ a      b      c
+     │ Int64  Int64  Int64
+─────┼─────────────────────
+   1 │     1      6     11
+   2 │     2      7     12
+   3 │     3      8     13
+   4 │     4      9     14
+   5 │     5     10     15
+"""
+function Base.invpermute!(df::AbstractDataFrame, p::AbstractVector{<:Integer})
```

Comment:
We could avoid some code duplication by using an internal `_permute!(invpermute!!, df, p)` which would also be used for `_permute!(permute!!, df, p)`.

Reply:
done